src/validation/json-schemas/collections/collections-write.json

```diff
@@ -68,28 +68,86 @@
         "dateCreated",
         "dataFormat"
       ],
-      "oneOf": [
-        {
-          "required": [
-            "parentId",
-            "contextId"
-          ]
-        },
+      "allOf": [
         {
           "allOf": [
             {
               "not": {
-                "required": [
-                  "parentId"
+                "allOf": [
+                  {
+                    "not": {
+                      "required": [
+                        "published"
+                      ]
+                    }
+                  },
+                  {
+                    "required": [
+                      "datePublished"
+                    ]
+                  }
                 ]
               }
             },
             {
-              "not": {
-                "required": [
-                  "contextId"
-                ]
-              }
+              "anyOf": [
+                {
+                  "properties": {
+                    "published": {
+                      "type": "boolean",
+                      "enum": [
+                        true
+                      ]
+                    }
+                  },
+                  "required": [
+                    "datePublished"
+                  ]
+                },
+                {
+                  "properties": {
+                    "published": {
+                      "type": "boolean",
+                      "enum": [
+                        false
+                      ]
+                    }
+                  },
+                  "not": {
+                    "required": [
+                      "datePublished"
+                    ]
+                  }
+                }
+              ]
+            }
+          ]
+        },
+        {
+          "oneOf": [
+            {
+              "required": [
+                "parentId",
+                "contextId"
+              ]
+            },
+            {
+              "allOf": [
+                {
+                  "not": {
+                    "required": [
+                      "parentId"
+                    ]
+                  }
+                },
+                {
+                  "not": {
+                    "required": [
+                      "contextId"
+                    ]
+                  }
+                }
+              ]
             }
```

Comment:
Consider this might be more readable as it puts all 3 conditions under the `anyOf` clause:

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://identity.foundation/dwn/json-schemas/collections-write.json",
  "type": "object",
  "additionalProperties": false,
  "required": [
    "authorization",
    "descriptor"
  ],
  "properties": {
    "authorization": {
      "$ref": "https://identity.foundation/dwn/json-schemas/general-jws.json"
    },
    "descriptor": {
      "type": "object",
      "properties": {
        "method": {
          "enum": [
            "CollectionsWrite"
          ],
          "type": "string"
        },
        "target": {
          "$ref": "https://identity.foundation/dwn/json-schemas/defs.json#/definitions/did"
        },
        "recipient": {
          "$ref": "https://identity.foundation/dwn/json-schemas/defs.json#/definitions/did"
        },
        "protocol": {
          "type": "string"
        },
        "contextId": {
          "type": "string"
        },
        "schema": {
          "type": "string"
        },
        "recordId": {
          "$ref": "https://identity.foundation/dwn/json-schemas/defs.json#/definitions/uuid"
        },
        "parentId": {
          "$ref": "https://identity.foundation/dwn/json-schemas/defs.json#/definitions/uuid"
        },
        "dataCid": {
          "type": "string"
        },
        "dateCreated": {
          "type": "number",
          "minimum": 0
        },
        "dataFormat": {
          "type": "string"
        }
      },
      "additionalProperties": false,
      "required": [
        "target",
        "method",
        "recordId",
        "dataCid",
        "dateCreated",
        "dataFormat"
      ],
      "allOf": [
        {
          "anyOf": [
            {
              "$comment": "rule defining relationships between `published` and `datePublished`"
            },
            {
              "properties": {
                "published": {
                  "type": "boolean",
                  "enum": [
                    true
                  ]
                },
                "datePublished": {
                  "type": "number",
                  "minimum": 0
                }
              },
              "required": [
                "datePublished"
              ]
            },
            {
              "properties": {
                "published": {
                  "type": "boolean",
                  "enum": [
                    false
                  ]
                }
              },
              "not": {
                "required": [
                  "datePublished"
                ]
              }
            },
            {
              "allOf": [
                {
                  "not": {
                    "required": [
                      "datePublished"
                    ]
                  }
                },
                {
                  "not": {
                    "required": [
                      "published"
                    ]
                  }
                }
              ]
            }
          ]
        },
        {
          "oneOf": [
            {
              "required": [
                "parentId",
                "contextId"
              ]
            },
            {
              "allOf": [
                {
                  "not": {
                    "required": [
                      "parentId"
                    ]
                  }
                },
                {
                  "not": {
                    "required": [
                      "contextId"
                    ]
                  }
                }
              ]
            }
          ]
        }
      ]
    },
    "encodedData": {
      "$ref": "https://identity.foundation/dwn/json-schemas/defs.json#/definitions/base64url"
    }
  }
}
```

Reply:
This is much cleaner. I've updated to be similar. Some of these block edge cases evaluating to true break my brain a bit :wink: